### PR TITLE
Building Python wheels for releases

### PR DESCRIPTION
**Pull request by thijsmie**
After much trial and error I managed to get the wheels rolling. This will build Cyclone DDS binaries ready to use for most common platforms (Linux x86, x64, Windows win32/x64, macOS x86_64/arm64), package the resulting library with the Python package in a `.whl` file and run the full Python testsuite against each one, apart from the macOS arm64 builds (there are no Apple Silicon agents on Azure Pipelines).

Each platform has its own config file. This leads to some duplication, but ultimately I found that getting it to work otherwise was a pain.

To get it to work, also some internals had to change:
 * Top level C Extension `ddspy` was renamed to `cyclonedds._clayer` to get `delocate` working.
 * The code that locates the ddsc lib was split up and extended to allow unit testing and in-wheel loading.
 * The pytest-mock dependency is introduced to mock python os functions to unit test loaders. This might come in handy with testing other things too.
 * Error handling in the C extension in case things are not loading to turn a segfault into an error if you mess up something while developing.

### CODE REVIEW

**Comment by thijsmie** [comment]
A successful wheel build run can be viewed [here](https://dev.azure.com/tmiedema/cyclonedds-python/_build/results?buildId=312&view=results).

**Review by eboasson — COMMENTED** [review]
It looks ok to me, and it works, so it has my blessing in principle.

The one thing that I am a little concerned about is the version number:
* @k0ekk0ek was having doubts about using the same version number for all components. I am fine either way, in my view whatever causes the least confusion for users is best. The trouble is that I don't know what that is, but we need to decide before bumping the version here. @k0ekk0ek, if you too can't choose, perhaps you can simply flip a coin?
* We can only do an official release after approval from Eclipse, but we can provide what the handbook calls "milestone builds". Those need to clearly state that they are not official releases. As a wheel involves making binaries available, would it be possible to make it version 0.8.0rc1 (or whatever the number will end up being) instead? That would definitely be the easiest way to deal with that detail.

**Comment by thijsmie** [comment]
I don't actually care what the version number ends up being, and merging this will not automatically generate a release, that only happens on a tag, so I can also just set it back to 0.0.0 and leave the discussion for later.

A bit about versioning in Python: there is of course a [PEP](https://www.python.org/dev/peps/pep-0440/) about it. If we end up using a `x.y.zrc1` scheme people will have to explicitly state the version while installing, since pip does not install release candidates by default. I think `pip install cyclonedds==0.8.0rc1` makes very clear no official release is being installed which hopefully should meet the Eclipse "milestone build" release requirements?

**Comment by k0ekk0ek** [comment]
It works, so that's a big plus :slightly_smiling_face: There's some things I want to have a better understanding off before I sign-off on them and there's things I think we should change before releasing. e.g. the `author_email` is still Thijs's ADLINK address. Which is accurate and he definitely should get the kudos, but we may want to choose something different there? Perhaps we can supply multiple, or the development mailing list. I don't know exactly what the right choice is yet, but I do want to form an opinion on the matter before pushing the first release. Of course, the project lead may overrule if he sees fit :wink:

As far as version numbering goes. It is my preference to not tie them together so that we can have different life cycles. e.g. we may want to do a 0.x.1 release for cyclonedds-python if desired without doing a cyclonedds-cxx patch release. Whether or not we should release a 0.x.0 for all products (semi-)simultaneously is another topic. My reasoning here is that I don't like to imply a significance in numbering where there is none. Or at least none that's backed by an automatic process that ensures correctness. I think using the same major/minor implies a false significance to users, but I may be wrong. On the other hand, I have no idea how the https://projects.eclipse.org/projects/iot.cyclonedds should/will look if we use different numbers. I do know that the "Eclipse" project itself ships out a product that has a specific version while each component has a version of their own so it might be that we want to ship Eclipse Cyclone DDS 0.8.0, which in turn consists of cyclonedds 0.8.0, cyclonedds-cxx x.x.0 and cyclonedds-python x.x.0. Eclipse Cyclone DDS 0.8.1 for instance can then consist of cyclonedds 0.8.0 (because it's perfect :sweat_smile:), cyclonedds-cxx x.x.1 and cyclonedds-python x.x.2.

**Review by k0ekk0ek — CHANGES_REQUESTED** [review]

File: src/cyclonedds/cyclonedds/internal.py
Comment (thijsmie) [reply src/cyclonedds/cyclonedds/internal.py]: This is now checked.

File: src/cyclonedds/cyclonedds/internal.py
Comment (thijsmie) [reply src/cyclonedds/cyclonedds/internal.py:46]: With name mangling in play this check is sometimes necessary.

**Pull request merged**
